lib/liquid_voting/delegations.ex

```diff
@@ -185,19 +186,47 @@ defmodule LiquidVoting.Delegations do
   def upsert_delegation(%{delegator_id: delegator_id, delegate_id: delegate_id} = attrs) do
     proposal_url = Map.get(attrs, :proposal_url)
 
-    Delegation
-    |> where(delegator_id: ^delegator_id)
-    |> Repo.all()
-    |> resolve_conflicts(delegate_id, proposal_url)
-    |> case do
-      {:ok, delegations} ->
-        delegations
-        |> find_similar_delegation_or_return_new_struct(proposal_url)
-        |> Delegation.changeset(attrs)
-        |> Repo.insert_or_update()
-
-      {:error, %{message: message, details: details}} ->
-        {:error, %{message: message, details: details}}
+    with {:ok} <- check_vote_conflict(delegator_id, proposal_url) do
+      Delegation
+      |> where(delegator_id: ^delegator_id)
+      |> Repo.all()
+      |> resolve_conflicts(delegate_id, proposal_url)
+      |> case do
+        {:ok, delegations} ->
+          delegations
+          |> find_similar_delegation_or_return_new_struct(proposal_url)
+          |> Delegation.changeset(attrs)
+          |> Repo.insert_or_update()
+
+        {:error, %{message: message, details: details}} ->
+          {:error, %{message: message, details: details}}
+      end
+    end
+  end
+
+  # Checks for a conflicting vote, in the case where a proposal-specific delegation creation is attempted.
+  #
+  # Returns {:ok} if delegation creation is for a global delegation, or if no conflicting vote is found.
+  # Returns an error, if a conflicting vote is found.
+  defp check_vote_conflict(delegator_id, proposal_url) do
+    if proposal_url != nil do
```

Comment:
Although I managed to avoid nested case statements in `upsert_delegation/1`, I used an `if` statement, which contains a nested case statement, in `check_vote_conflict/2`. This seems inelegant and unnecessarily opaque. Any ideas to improve this approach would be gratefully received.